pages/signup.tsx

```diff
@@ -0,0 +1,384 @@
+    const router = useRouter();
+    const [currStage, setCurrStage] = useState(0);
+    const [email, setEmail] = useState("");
+    const [emailInvalid, setEmailInvalid] = useState({
+        isInvalid: false,
+        reason: "",
+    });
+    const [name, setName] = useState("");
+    const [password, setPassword] = useState("");
+    const [passwordInvalid, setPasswordInvalid] = useState({
+        isInvalid: false,
+        reason: "",
+    });
+    const [termsAgreement, setTermsAgreement] = useState({
+        saveEmail: false,
+        saveProgress: false,
+        participateSurvey: false,
+    });
+    const [canContinue, setCanContinue] = useState(false);
+    const [fromGoogle, setFromGoogle] = useState(false);
+
+    useEffect(() => {
+        // Go directly to the TOS if coming from the google flow
+        if (!router.isReady) {
+            // router not ready yet
+            // TODO: add spinner
+        }
+        if (router.query["GoogleFlow"]) {
+            setFromGoogle(true);
+            setCurrStage(3);
+        } else {
+            setFromGoogle(false);
+        }
+    }, [router.isReady]);
+
+    const userExist = async (email) => {
+        return await Auth.signIn(email.toLowerCase(), "123")
+            .then((res) => {
+                return false;
+            })
+            .catch((error) => {
+                const code = error.code;
+                switch (code) {
+                    case "UserNotFoundException":
+                        return false;
+                    case "NotAuthorizedException":
+                        return true;
+                    case "PasswordResetRequiredException":
+                        return true;
+                    default:
+                        return false;
+                }
+            });
+    };
+
+    const validateEmail = async () => {
+        if (email == "") {
+            // don't set as error state if it's currently empty
+            setEmailInvalid({ isInvalid: false, reason: "" });
+            setCanContinue(false);
+        } else if (isEmail(email)) {
+            // Check if account already exists
+            const emailExists = await userExist(email);
+            if (emailExists) {
+                setEmailInvalid({
+                    isInvalid: true,
+                    reason: "Account already exists with this email",
+                });
+                setCanContinue(false);
+            } else {
+                setEmailInvalid({ isInvalid: false, reason: "" });
+                setCanContinue(true);
+            }
+        } else {
+            setEmailInvalid({
+                isInvalid: true,
+                reason: "Not a valid email format",
+            });
+            setCanContinue(false);
+        }
+    };
+
+    const validateName = () => {
+        if (name == "") {
+            setCanContinue(false);
+        } else {
+            setCanContinue(true);
+        }
+    };
+
+    const validatePassword = () => {
```

Comment:
lets move this into a helper function so it can be used in both reset and signup